src/v/cloud_storage_clients/s3_client.cc

```diff
@@ -706,34 +683,59 @@ ss::future<s3_client::list_bucket_result> s3_client::do_list_objects_v2(
     auto header = _requestor.make_list_objects_v2_request(
       name,
       std::move(prefix),
       std::move(start_after),
       max_keys,
-      continuation_token);
+      std::move(continuation_token),
+      delimiter);
     if (!header) {
         return ss::make_exception_future<list_bucket_result>(
           std::system_error(header.error()));
     }
     vlog(s3_log.trace, "send https request:\n{}", header.value());
     return _client.request(std::move(header.value()), timeout)
-      .then([](const http::client::response_stream_ref& resp) mutable {
-          return util::drain_chunked_response_stream(resp).then(
-            [resp](iobuf outbuf) {
+      .then([pred = std::move(collect_item_if)](
+              const http::client::response_stream_ref& resp) mutable {
+          return resp->prefetch_headers().then(
+            [resp, gather_item_if = std::move(pred)]() mutable {
                 const auto& header = resp->get_headers();
                 if (header.result() != boost::beast::http::status::ok) {
-                    // We received error response so the outbuf contains
-                    // error digest instead of the result of the query
                     vlog(s3_log.warn, "S3 replied with error: {:l}", header);
-                    return parse_rest_error_response<
-                      s3_client::list_bucket_result>(
-                      header.result(), std::move(outbuf));
+
+                    // In the error path we drain the response stream fully, the
+                    // error response should not be very large.
+                    return util::drain_chunked_response_stream(resp).then(
+                      [result = header.result()](iobuf buf) {
+                          return parse_rest_error_response<list_bucket_result>(
+                            result, std::move(buf));
+                      });
                 }
-                auto res = iobuf_to_list_bucket_result(std::move(outbuf));
-                return ss::make_ready_future<list_bucket_result>(
-                  std::move(res));
+
+                return ss::do_with(
+                  resp->as_input_stream(),
+                  xml_sax_parser{},
+                  [pred = std::move(gather_item_if)](
+                    ss::input_stream<char>& stream, xml_sax_parser& p) mutable {
+                      p.start_parse(
+                        std::make_unique<aws_parse_impl>(std::move(pred)));
+                      return ss::do_until(
+                               [&stream] { return stream.eof(); },
+                               [&stream, &p] {
+                                   return stream.read().then(
+                                     [&p](ss::temporary_buffer<char>&& chunk) {
+                                         p.parse_chunk(std::move(chunk));
+                                     });
+                               })
+                        .then([&p] {
+                            p.end_parse();
```

Comment:
From memory of issues with input_stream in the past, I think we need to call `stream.close()` here, or there is a risk of asserting in the seastar reactor if there was a readahead I/O in flight when an exception is thrown from the body of our do_until.

Both here and in the equivalent ABS function.

Reply:
sure, added `stream.close()` in both places